src/mainloop.rkt

```diff
@@ -328,24 +328,26 @@
   (define original-points (setup-context! vars (or specification prog) precondition repr))
   (run-improve! iters prog specification preprocess original-points repr))
 
-
 (define (make-preprocess-pcontext prog pcontext iters
                                   #:specification [specification #f]
                                   #:preprocess [preprocess '()])
   (timeline-event! 'preprocess)
 
   ;; If the specification is given, it is used for sampling points
-  (define sortable (connected-components specification (*context*)))
-
-  (define new-preprocess
-    (for/list ([sortable-variables (in-list sortable)]
+  (define preprocess-sort
+    (for/list ([sortable-variables (connected-components specification (*context*))]
                #:when (> (length sortable-variables) 1))
       (cons 'sort sortable-variables)))
-  (timeline-push! 'symmetry (map ~a new-preprocess))
-  (*herbie-preprocess* (append preprocess new-preprocess))
+  (define preprocess-abs
+    (for/list ([absable-variable (get-fabs specification (*context*))])
+      (list 'abs absable-variable)))
```

Comment:
This is OK but note that both `sort` and `abs` preprocessing steps basically involve using an e-graph to make equality queries. I think we will actually want to do one egraph query that does both; that is, if you have an expression `f(x, y)`, you'll want to create an egraph with nodes `f(x, y)`, `f(y, x)`, `f(-x, y)`, and `f(x, -y)`, and see which ones all end up equal, and then use that to determine what preprocessing to do.

Doing one query should be faster and more extensible than doing one query for every kind of preprocessing step.